### PR TITLE
Display Name Support for derived RecordTypes

### PR DESCRIPTION
**Pull request by lesaltzm**

### CODE REVIEW

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:x: 7 Binary Compatibility issues for **Microsoft.PowerFx.Core** compared against [version 0.2.2-preview.20220818-317600](https://www.nuget.org/packages/Microsoft.PowerFx.Core/0.2.2-preview.20220818-317600).
<details>

```
CannotSealType : Type 'Microsoft.PowerFx.Types.AggregateType' is effectively (has a private constructor) sealed in the implementation but not sealed in the contract.
MembersMustExist : Member 'Microsoft.PowerFx.Types.AggregateType..ctor(System.Boolean)' does not exist in the implementation but it does exist in the contract.
CannotMakeMemberNonVirtual : Member 'Microsoft.PowerFx.Types.AggregateType.FieldNames' is non-virtual in the implementation but is virtual in the contract.
CannotMakeMemberNonVirtual : Member 'Microsoft.PowerFx.Types.AggregateType.Equals(System.Object)' is non-virtual in the implementation but is virtual in the contract.
CannotMakeMemberNonVirtual : Member 'Microsoft.PowerFx.Types.AggregateType.FieldNames.get()' is non-virtual in the implementation but is virtual in the contract.
CannotMakeMemberNonVirtual : Member 'Microsoft.PowerFx.Types.AggregateType.GetHashCode()' is non-virtual in the implementation but is virtual in the contract.
CannotMakeMemberNonVirtual : Member 'Microsoft.PowerFx.Types.AggregateType.TryGetFieldType(System.String, Microsoft.PowerFx.Types.FormulaType)' is non-virtual in the implementation but is virtual in the contract.
```

</details>
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

File: src/tests/Microsoft.PowerFx.Core.Tests/TypesAndDisplayNameTests/LazyTypeDisplayNameTests.cs [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+                    "Foo" => "FooDisplay",
+                    "Bar" => "BarDisplay",
+                    "Baz" => "BazDisplay",
+                    _ => null
+                };
+
+                return displayName != null;
+            }
+
+            public override bool TryGetLogicalName(string displayName, out string logicalName)
```
Comment (MikeStall) [inline src/tests/Microsoft.PowerFx.Core.Tests/TypesAndDisplayNameTests/LazyTypeDisplayNameTests.cs:52]: >TryGetLogicalName [](http://example.com/codeflow?start=33&length=17)

As a sanity check -  
can we also emulate TryGetLogicalName via  FieldNames + TryGetDisplayName? 

Put another way, could we have a reasonable default implementation ?

Comment (lesaltzm) [reply src/tests/Microsoft.PowerFx.Core.Tests/TypesAndDisplayNameTests/LazyTypeDisplayNameTests.cs:52]: Yes, we could iterate FieldNames , and call `TryGetDisplayName` for each, stopping when we find a match. 
Not sure if it's worth having a slow default implementation vs asking hosts to override both.

**Comment by lesaltzm** [comment]
> CannotSealType : Type 'Microsoft.PowerFx.Types.AggregateType' is effectively (has a private constructor) sealed in the implementation but not sealed in the contract.
MembersMustExist : Member 'Microsoft.PowerFx.Types.AggregateType..ctor(System.Boolean)' does not exist in the implementation but it does exist in the contract.
CannotMakeMemberNonVirtual : Member 'Microsoft.PowerFx.Types.AggregateType.FieldNames' is non-virtual in the implementation but is virtual in the contract.
CannotMakeMemberNonVirtual : Member 'Microsoft.PowerFx.Types.AggregateType.Equals(System.Object)' is non-virtual in the implementation but is virtual in the contract.
CannotMakeMemberNonVirtual : Member 'Microsoft.PowerFx.Types.AggregateType.FieldNames.get()' is non-virtual in the implementation but is virtual in the contract.
CannotMakeMemberNonVirtual : Member 'Microsoft.PowerFx.Types.AggregateType.GetHashCode()' is non-virtual in the implementation but is virtual in the contract.
CannotMakeMemberNonVirtual : Member 'Microsoft.PowerFx.Types.AggregateType.TryGetFieldType(System.String, Microsoft.PowerFx.Types.FormulaType)' is non-virtual in the implementation but is virtual in the contract.


This seems odd. The only intended breaking change here is 1 and 2 (making the constructor for AggregateType non-public). The other members remain virtual (I double checked), I wonder if this might be a bug in the compat checker.

**Comment by lesaltzm** [comment]
Closed, decided exposing DisplayNameProvider directly may be more useful.

**Pull request closed without merging**